pkg/ibm/ibmz_helpers.go

```diff
@@ -287,3 +290,41 @@ func (iz IBMZDynamicConfig) assignIpToInstance(instance *vpcv1.Instance, vpcServ
 	ip, err = assignNewlyAllocatedIP(instance, vpcService)
 	return ip, err
 }
+
+// findInstancesWithoutTaskRuns iterates over instances retrieved from the iz cloud and returns a list of those that
+// are associated with a non-existing Tekton TaskRun. Each instance's volume should have a tag with the associated
+// TaskRun's namespace and name. This is compared to existingTaskRuns, which is a map of namespaces to a list of
+// TaskRuns in that namespace, to determine if this instance's TaskRun still exists.
+func (iz IBMZDynamicConfig) findInstancesWithoutTaskRuns(log logr.Logger, vpcService *vpcv1.VpcV1, instances []vpcv1.Instance, existingTaskRuns map[string][]string) []string {
+	var instancesWithoutTaskRuns []string
+
+	// Iterate over all VM instances
+	for _, instance := range instances {
+		volumeId := instance.VolumeAttachments[0].ID
+		// Get instance's volume; assumes only one volume per instance
+		volume, _, err := vpcService.GetVolume(&vpcv1.GetVolumeOptions{ID: volumeId})
+		if err != nil {
+			msg := fmt.Sprintf("failed to get instance %s's volume; skipping this instance...", *instance.ID)
+			log.Error(err, msg)
+			continue
+		}
+		// Iterate over the volume's tags
+		for _, volumeTag := range volume.UserTags {
+			err := cloud.ValidateTaskRunId(volumeTag)
+			if err != nil {
+				log.Error(err, "invalid TaskRun ID; appending to list anyway...", "instanceID", *instance.ID)
+				break
+			}
```

Comment:
this is not clear to me. I read it as take the first volume's userTag and if it's not a ValidTaskRunId skip to the next instance. Am I right?